src/core/lib/slice/slice_buffer.cc

```diff
@@ -87,13 +87,18 @@ void grpc_slice_buffer_destroy(grpc_slice_buffer* sb) {
   }
 }
 
-uint8_t* grpc_slice_buffer_tiny_add(grpc_slice_buffer* sb, size_t n) {
+template <bool maybe_empty>
+static uint8_t* tiny_add_internal(grpc_slice_buffer* sb, size_t n) {
```

Comment:
did you try changing the code to something like the following?  compiler should be able to merge the "sb->count == 0" branches and with that you won't need a template param.
```
maybe_embiggen(sb);
if (sb->count == 0) goto add_first;
```


if (back.refcount) {
  ++back;
  back.refcount = nullptr;
}

Also note that when count is 0, calling maybe_embigen can (and probably does) make a difference because we would potentially reset the slices pointer and save ourselves some allocations.  So, probably pulling it up the stack is going to help too.

Reply:
Not yet, but will give it a shot.

For hpack_enc, however, count will demonstrably never be 0. That's the main reason for the template parameter, to save that branch.

For now, I'm happy to split this into two PRs; one which does the coalescing sans API change, which I think is a must do - and the other with the API change. That way we can also tease out whether the API change is worth the change in specific performance or not.

Reply:
sg.